File: 백준/Silver/1463. 1로 만들기/1로 만들기.js
const fs = require('fs');
// const input = fs.readFileSync('example.txt').toString().trim();
const input = fs.readFileSync('/dev/stdin').toString().trim();

n = +input;

let dp = [0, 0, 1];

for(let i = 3; i < n+1; i++){
    if(i%2 === 0 && i%3 === 0){
        dp[i] = Math.min(dp[i/3]+1, dp[i/2]+1, dp[i-1]+1);
    } else if(i%3 === 0){
        dp[i] = Math.min(dp[i/3]+1, dp[i-1]+1);
    } else if(i%2 === 0){
        dp[i] = Math.min(dp[i/2]+1, dp[i-1]+1);
    } else{
        dp[i] = dp[i-1]+1;
    }
}

console.log(dp[n]);

File: 백준/Silver/1463. 1로 만들기/1로 만들기.py
n=int(input())
dp=[0]*(n+2)
dp[0]=0 #n==0
dp[1]=0 #n==1
dp[2]=1 #n==2

def cul(a):
    if dp[a]==0:
        if a%2==0 and a%3==0:
            dp[a]=min(dp[a//3]+1,dp[a-1]+1,dp[a//2]+1)
        elif a%3==0:
            dp[a]=min(dp[a//3]+1,dp[a-1]+1)
        elif a%2==0:
            dp[a]=min(dp[a//2]+1,dp[a-1]+1)
        else:
            dp[a]=dp[a-1]+1

for i in range(3,n+1):
    cul(i)

print(dp[n])




